pkg/block/fetcher.go

```diff
@@ -0,0 +1,410 @@
+package block
+
+import (
+	"context"
+	"encoding/json"
+	stderrors "errors"
+	"io/ioutil"
+	"os"
+	"path"
+	"path/filepath"
+	"sync"
+	"time"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/oklog/ulid"
+	"github.com/pkg/errors"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/prometheus/pkg/labels"
+	"github.com/prometheus/prometheus/pkg/relabel"
+	tsdberrors "github.com/prometheus/prometheus/tsdb/errors"
+	"github.com/prometheus/prometheus/tsdb/fileutil"
+	"github.com/thanos-io/thanos/pkg/block/metadata"
+	"github.com/thanos-io/thanos/pkg/extprom"
+	"github.com/thanos-io/thanos/pkg/model"
+	"github.com/thanos-io/thanos/pkg/objstore"
+	"github.com/thanos-io/thanos/pkg/runutil"
+)
+
+type syncMetrics struct {
+	syncs        prometheus.Counter
+	syncFailures prometheus.Counter
+	syncDuration prometheus.Histogram
+
+	synced *extprom.TxGaugeVec
+}
+
+const (
+	syncMetricSubSys = "blocks_meta"
+
+	corruptedMeta = "corrupted-meta-json"
+	noMeta        = "no-meta-json"
+	loadedMeta    = "loaded"
+	failedMeta    = "failed"
+
+	// Filter's label values.
+	labelExcludedMeta = "label-excluded"
+	timeExcludedMeta  = "time-excluded"
+	TooFreshMeta      = "too-fresh"
```

Comment:
Does this need to be exposed?

Reply:
Yes - used in https://github.com/thanos-io/thanos/pull/1937 (: